modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,705 @@
+
+    :throws: :class:`OffsetNotFoundError` if a valid byte offset cannot be found
+             in any of the chunks
+
+    See :proc:`~IO.open` for other errors that could be thrown when attempting
+    to open the file
+  */
+  proc readParallelLocal(filePath: string, type t, in delim: ?dt = b"\n",
+                         header = headerPolicy.noHeader, nTasks: int = here.maxTaskPar,
+                         type deserializerType = defaultDeserializer
+  ): [] t throws
+    where dt == bytes || dt == string
+  {
+    var results: [0..nTasks] list(t);
+    const f = open(filePath, ioMode.r),
+          fileBounds = 0..<f.size,
+          byteOffsets = findDelimChunksChecked(f, delim, nTasks, t, fileBounds,
+                                               header, deserializerType);
+
+    coforall tid in 0..<nTasks with (ref results) {
+      const taskBounds = byteOffsets[tid]..byteOffsets[tid+1];
+      var des: deserializerType,
+          r = f.reader(locking=false, region=taskBounds, deserializer=des),
+          s = new t();
+
+      // read all the 't' values in the chunk into a list
+      while r.read(s) do
+        results[tid].pushBack(s);
+    }
+
+    // gather results from each task into an array
+    const nPerTask = [r in results] r.size,
+          itemOffsets = (+ scan nPerTask) - nPerTask;
+
+    var result: [0..<itemOffsets.last] t;
+    coforall tid in 0..<nTasks do
+      result[itemOffsets[tid]..<itemOffsets[tid+1]] = results[tid].toArray();
+
+    return result;
+  }
+
+  /*
+    Get an array of ``n+1`` byte offsets that divide the file ``f`` into ``n``
+    roughly equally sized chunks, where each byte offset lines up with a
+    delimiter.
+
+    :arg f: the file to search
+    :arg delim: the delimiter to use to separate the file into chunks
+    :arg n: the number of chunks to find
+    :arg bounds: a range of byte offsets to break into chunks
+    :arg header: a header policy to use when searching for the first byte offset
+
+    :returns: a length ``n+1`` array of byte offsets (the last offset is
+              ``bounds.high``)
+
+    :throws: :class:`OffsetNotFoundError` if a valid byte offset cannot be found
+              in any of the chunks
+  */
+  proc findDelimChunks(const ref f: file, in delim: ?dt, n: int, bounds: range,
+                       header = headerPolicy.noHeader
+  ): [] int throws
+    where dt == bytes || dt == string
+  {
+    const (findStart, startOffset) = header.apply(f, bounds),
+          chunkIndices = if findStart then 0..<n else 1..<n;
+
+    const nDataBytes = bounds.high - startOffset,
+          approxBytesPerChunk = nDataBytes / n;
+
+    var chunkOffsets: [0..n] int;
+    chunkOffsets[0] = startOffset;
+    chunkOffsets[n] = bounds.high;
+
+    coforall i in chunkIndices with (ref chunkOffsets) {
+      const estOffset = chunkOffsets[0] + i * approxBytesPerChunk,
+            r = f.reader(locking=false, region=estOffset..);
```

Comment:
Doesn't this skip the first line/delimited element?

You set `chunkOffsets[0]` earlier but htis loop will overwrite it. Worse, it will overwrite it in a racy way, won't it?

Reply:
It can skip the first element depending on the header type selected by the user and how the file is formatted. For example, for headerKind.noHeader, it assumes that an item can be read at `bounds.low`, so nothing is skipped. Alternatively, for `headerKind.find`, it looks for the first offset where a delimiter can be found (this could skip the first element if there isn't a delimiter before it, but `find` is more intended for use in `readParallel` than `readParallelDelimited`).

There is a bit of a race because the other tasks `estOffset` value could depend on whether they run before or after the 0th task. I'll fix that.